### ISSUE TITLE
Standardize data types / concept schemes for NACE code

### ISSUE THREAD

**Issue by pvdbosch** [belgif/fedvoc#30]
There are different variants. Each additional digit in the code is a more detailed subcategory.
* European NACE codes
  * https://ec.europa.eu/eurostat/web/nace/overview
  * [V2](https://showvoc.op.europa.eu/#/datasets/ESTAT_Statistical_Classification_of_Economic_Activities_in_the_European_Community_Rev._2/data?resId=http:%2F%2Fdata.europa.eu%2Fux2%2Fnace2%2F5224) and [V2.1](https://showvoc.op.europa.eu/#/datasets/ESTAT_Statistical_Classification_of_Economic_Activities_in_the_European_Community_Rev._2.1._%28NACE_2.1%29/data)
  * up to four digits, formatted xx.yy
* Belgian NACE codes which are a superset (specialization) of european ones
  * up to seven digits
  * [FPS eco](https://economie.fgov.be/en/themes/enterprises/crossroads-bank-enterprises/services-administrations/tables-codes)
  * NSSO seems to use less digits (x>=1000 x<=99999) - [See Annex 31](https://www.socialsecurity.be/lambda/portail/glossaires/bijlagen.nsf/web/Bijlagen_Home_fr)
  * versions: Nace2003 - Nace2008

**Comment by pvdbosch** [comment]
From [what I can find](https://www.vlaio.be/nl/begeleiding-advies/start/opstart-van-je-onderneming/het-belang-van-correcte-nace-codes), there are three main contexts in which NACE-codes are used in Belgium:
* VAT (FPS Fin): at organization (enterprise) level => 5 digit version (TBC)
* ONSS: at organization + establishment level  => 5 digit version
* economical activities (CBE) : at establishment level => 2-7 variable length 

Examples:
* [on enterprise](https://kbopub.economie.fgov.be/kbopub/toonondernemingps.html?ondernemingsnummer=247499953) : NSSO + vat
* [on establishment](https://kbopub.economie.fgov.be/kbopub/toonvestigingps.html?vestigingsnummer=2123471342) : NSSO + CBE

The different variants could be represented like this:

```YAML
NaceCode:
  description: NACE activity code (international). First two digits are division of activity. Each further digits indicates an additional level of detail within the category of the previous level. The meaning of the first four digits are standard across all countries. Further digits may vary from country to country.
  type: string
  pattern: '^\d{2}\d*$'

NaceBel2008Code:
  description: Belgian NACE activity code. The meaning of the first four digits are standard across all countries. The last 3 digits are defined in the Belgian NACE 2008 list.
  type: string
  pattern: '^\d{2}\d{0,5}$'

NaceBel2008Length5Code:
  description: Belgian NACE activity code of length 5, as used for NSSO and for VAT purposes. The meaning of the first four digits are standard across all countries. The last digit is defined in the Belgian NACE 2008 list.
  type: string
  pattern: '^\d{5}$'
```

We could also use allOf to indicate a specialization (NaceCode > NaceBelCode > NaceBel5Code), TBD what the impact would be.
More specialized types are less flexible to change, but more clear in which detail level is used and more useful for input validation.

**Comment by pvdbosch** [comment]
I tweaked above openapi propositions a bit: 
* specified "of 2 up to 7 digits long" in description of NaceBel2008Code
* added comments to NaceBel2008Code and NaceBel2008Length5Code to indicate their relationship. I wouldn't put this in the schema description, bc its for authors of an OpenAPI rather than its users (clients).

These types can be added to a new openapi schema belgif-openapi-organization.

**Comment by MarcBruyland** [comment]
Closed as decided in the fedvoc meeting of 2024-06-28.

**Comment by barthanssens** [comment]
FWIW there has been a NACE2008 skos thesaurus (https://vocab.belgif.be/auth/nace2008?lang=en) available for quite a few years now

**Comment by pvdbosch** [comment]
@barthanssens , would it be best then to use http://vocab.belgif.be/auth/nace2008 as the URI of the ConceptScheme? 

A link could also be added in the comments so users can find the URIs of the individual Nace2008 concepts.

So in NaceBel2008 ConceptScheme of fedvoc:
* URI:<http://vocab.belgif.be/ns/other#nacebel2008#id> => <http://vocab.belgif.be/auth/nace2008>
* Comments: See https://statbel.fgov.be/nl/over-statbel/methodologie/classificaties/nace-bel-2008 **and http://vocab.belgif.be/auth/nace2008**

**Comment by barthanssens** [comment]
The URI for the skos:ConceptScheme is indeed http://vocab.belgif.be/auth/nace2008  
 (not http://vocab.belgif.be/ns/other#nacebel2008#id)

**Comment by pvdbosch** [comment]
OK, would be best to use the existing URI then. @MarcBruyland ,  could the changes in my previous command by applied to fedvoc?

**Comment by pvdbosch** [comment]
For the 2025 codes, from what I can find, they will only be up to 5 digits long:

* https://statbel.fgov.be/nl/over-statbel/methodologie/classificaties/nacebel-2025
* https://www.moore.be/nl/nieuws/twee-belangrijke-wijzigingen-voor-bedrijven-vanaf-1-januari-2025

Existing NaceCode notation could still be used for international scheme.
A NaceBel2025Code could be introduced for variable length 2-5
A NaceBel2008Length5Code could be introduced for fixed length 5.

An another node, do we need to extend NaceBel2025Code and NaceCode to include the top-level single letter categories?
They also can be found in the categorization on KBO e.g. https://kbopub.economie.fgov.be/kbopub/naceToelichting.html?nace.code=J , but not directly associated to enterprises.

**Comment by pvdbosch** [comment]
The top level NACE code of one letter (A-Z) has been confirmed to be used by FPS Economy.

So suggested to allow these also in the schemas:
```
NaceCode:
  description: NACE activity code (international). Either a single-letter code, or numeric code. A longer code specifies an additional level of detail within the category of the previous level. The first two digits are division of activity. The meaning of the single-letter codes and first four digits of a numeric code are standard and defined by EU. Further digits may vary from country to country.
  type: string
  pattern: '^([A-Z]{1}|\d{2}\d*)$'

NaceBel2008Code:
  description: Belgian NACE activity code of either a single letter, or numerical 2 up to 7 digits long. A longer code specifies an additional level of detail within the category of the previous level. The meaning of the single-letter codes and first four digits of a numeric code are standard across all countries. The last 3 digits are defined in the Belgian NACE 2008 list.
  type: string
  pattern: '^([A-Z]{1}|\d{5})$'
  # this is a more constrained version of the international NaceCode schema
```

About the length 7 codes: 
* they are defined in the list of FPS Economy: https://economie.fgov.be/nl/themas/ondernemingen/kruispuntbank-van/diensten-voor-administraties/codetabellen
* but not in the statbel list: https://statbel.fgov.be/nl/over-statbel/methodologie/classificaties/nace-bel-2008

On the NACE BEL 2025 codes:
* follows the same notation as 2008, but the codes themselves are backwards-incompatible (changed codes for same business meaning, new codes, split codes, ...); source: https://statbel.fgov.be/en/about-statbel/methodology/classifications/nacebel-2025
* There's no variant list with 7-digit codes published by FPS Economy available at this point. Maybe they won't maintain create one?
  * From this [source](https://www.moore.be/nl/nieuws/twee-belangrijke-wijzigingen-voor-bedrijven-vanaf-1-januari-2025):  De huidige zeven-cijferige codes, die dateren uit 2008, worden vervangen door vijfcijferige codes. De oude zeven-cijferige codes verdwijnen. Vanaf 1 januari 2025 zullen de btw en de RSZ de nieuwe structuur al hanteren. De Kruispuntbank van Ondernemingen (KBO) zal de wijziging later doorvoeren.
    * this web page isn't from a government instance however

I'd recommend to manage the 2025 codes in two separate lists then (ConceptSchemes, code tables) because they're functionally incompatible. Their value should never be used without knowing the version of the code list. 
For the notation, we could reuse the same data types, but an API should always specify which version of the code list is being used. 

If we decide to reuse the same schemas, we'd have to remove the "2008" part of the NaceBel2008Code and NaceBel2008Length5Code schemes.

**Comment by barthanssens** [comment]
Mixing openAPI schema's with semantic vocabularies here ...


I'm actually not in favor of adding datatypes to RDF (which is what FedVoc seems to be trying to do), feel free to do so in OpenAPI.

In practice, the OpenAPI equivalent of the datatype regex pattern would be a SHACL validation rule, not a datatype.
(yes, one can define custom datatypes in OWL2 with these kind of restrictions, but SHACL-engines are more common than OWL2-tools - RDF implementations are not required to support custom datatypes by the way - so in practice checking of the lexical value in an RDF-flow will be done in SHACL)

NACE codelist is (also) managed as as SKOS vocabulary (https://vocab.belgif.be/auth/nace2008), so there can be semantic links between different versions. There will be an updated version for 2025, with skos relations to the European NACE (if there would be a difference, otherwise it doesn't make sense to come up with a Belgian copy).

If someone wants to use a semantics, the URI of the SKOS Concept should be used  (see also the companyActivity property example in https://semiceu.github.io/Core-Business-Vocabulary/releases/2.2.0/#Code, which uses the EU NACE URI)

Also, a conceptScheme is not (necessarily) a datatype

**Comment by pvdbosch** [comment]
@barthanssens to enable better interoperability, there's a need to document standardized notations (i.e. encodings) of concepts, more broadly than only for REST.

Formerly this was documented as part of ConceptScheme, but later refactored into separate DataType entries in fedvoc to avoid conflating the two concerns. This was especially a problem when multiple notations exist. It was inspired on EU's use of data types for country notations (see https://github.com/belgif/fedvoc/issues/27#issuecomment-1735069532 ) .

Not being an expert in SHACL, is there an alternative way to do this without being constrained to a single application context?
If so, how would this look like, e.g. for country ISO alpha-2 and ISO num-3 codes?

**Comment by barthanssens** [comment]
Ok, got it @pvdbosch 

To be clear, you _can_ define a custom datatype, and check in SHACL a literal has that specific datatype. says `"ex:NaceBel2008"`
What is missing, is an easy way to check if all the literals with that datatype actually have a valid value

E.g. "bananas"^^ex:NaceBel2008` would be considered OK

(This works for 'built-in' XML Schema datatypes because libraries/stores can check them, but for custom data types all bets are off... 
It could be done when a validator supports  a custom SPARQL query, but that is not guaranteed to work, and would somewhat defeat the purpose of using SHACL instead of a series of custom scripts...)

So it boils down to defining `shacl:pattern` on properties (setting a shacl:pattern on a datatype/class is not defined in the spec, and e.g the online SHACL Playground tries to apply it to the name of the datatype instead of the value, so not going to work)

Meaning that one can say in SHACL that

- a property `ex:nace2008` must follow something like `sh:pattern "^\\d{5,7}$"` (and is of type `sh:datatype ex:NaceBel2008` if you really want to, but once again this will just check if the custom datatype is explicitly set, not if the value is correct)
- a property 'ex:alpha2' with `sh:pattern "^[A-Z]{2}$"`
- a property 'ex:alpha3' with `sh:pattern "^[A-Z]{3]$"`


(It becomes messy for generic properties like 'ex:countryCode', if both Alpha2 and Alpha3 would be allowed)

Alternatively, it is also possible to use SKOS, but in that case you need a thesaurus with all values, assign them a URI, and then check the format of the URI (see also https://github.com/SEMICeu/DCAT-AP/blob/master/releases/3.0.0/html/shacl/mdr-vocabularies.shape.ttl), or ADMS for more complex identifiers (e.g for KBO/CBE: to keep track that the company number is a number assigned by the company register etc).


To summarize: n the FedVoc abstraction layer it could be helpful to define datatypes, as long as one doesn't expect them to be 1-to-1 mapped to RDF datatypes (which now seems to be the case)

**Comment by pvdbosch** [comment]
@barthanssens , OK I propose to leave it as-is then for documentation purposes, without expecting these data types to be usable in this form to validate values when using Semantic Web tooling.

For REST services, the datatypes are being mapped 1-to-1 to JSON Schema types (for use in OpenAPI), but that's not part of the FedVoc.

> (It becomes messy for generic properties like 'ex:countryCode', if both Alpha2 and Alpha3 would be allowed)

That's indeed the problem we're facing; defining specific properties per datatype doesn't seem to be done a lot AFAIK.

From what I can find about adms:Identifier, the data types in FedVoc would correspond then to the identifierScheme prop, which seems to have a role as datatype of the identifier's content. From https://www.w3.org/TR/vocab-adms/#identifier : 

> the content string should be provided using [skos:notation](https://www.w3.org/TR/vocab-adms/#skos_notation) , datatyped with the identifier scheme (including the version number if appropriate);

**Comment by barthanssens** [comment]
> From what I can find about adms:Identifier, the data types in FedVoc would correspond then to the identifierScheme value, which can be used as datatype of the identifier's content. From https://www.w3.org/TR/vocab-adms/#identifier :
> 
> > the content string should be provided using [skos:notation](https://www.w3.org/TR/vocab-adms/#skos_notation) , datatyped with the identifier scheme (including the version number if appropriate);

@pvdbosch 

It used to be that way, but in ADMS v2 there is somewhat less emphasis on the datatyping (though it is still mentioned somewhere in the document)

https://semiceu.github.io/ADMS/releases/2.00/#Identifier

And using a datatype doesn't seem to be popular in practice, e.g. at least two (non-normative) examples where SEMIC / JRC are just using a string ...

- https://semiceu.github.io/Core-Business-Vocabulary/releases/2.2.0/#examples
- https://ec-jrc.github.io/dcat-ap-jrc/#ex-dataset

**Comment by pvdbosch** [comment]
Based on most recent info (e.g. [CBE publication](https://economie.fgov.be/sites/default/files/Files/Entreprises/KBO/codes-NACEBEL-2025.xlsx), we know that CBE will continue to use a NACEBEL 2025 variant of codes up to 7 digits, like they do for NACEBEL2008.
They also use the single-letter categories.

Based on this, I propose to 

1) add the NaceBel2025 ConceptScheme which extends the already defined NaceEuRev2_1 concept scheme.
 
2) for the data types, either 

a) unify them for 2008 and 2025:
* modify "NaceCode" to include single-letter codes
* rename NaceBel2008Code to NaceBelCode and include single letter codes
* rename NaceBel2008Length5Code to NaceBelLength5Code (pattern unchanged)

This expresses that the code lists are changed (new ConceptScheme), but the format doesn't.

b) or create separate data types for 2025:
 
* create "Nace2_0Code" and Nace2_1Code" replacing NaceCode and including single-letter codes
* create NaceBel2025Code, with same pattern as NaceBel2008Code (+ include single letter codes in both)
* create NaceBel2025Length5Code with same pattern as NaceBel2008Length5Code  (pattern unchanged)

For REST APIs, it documents more clearly which version is expected, bc the description of the schema is usually not overridden.

For Semantic Web tooling, it doesn't have much impact either way bc datatypes aren't validated.

For properties that may hold any version of the NACE Code, it's difficult to predict future evolution (e.g. NACE 2003 allowed two-letter codes), so only a string format without constraints can be used and needs to be accompanied with a field indicating which ConceptScheme is being used (or use a full URI in Semantic Web).

**Comment by pvdbosch** [comment]
FedVoc WG decided to create separate types for 2025.

Proposed schemas in https://github.com/belgif/openapi-organization/pull/11/files#diff-98b3f30bc09796079adfe096793b2997291790c85a008642dff33fc90a78eb5a

**Comment by MarcBruyland** [comment]
Closed as decided by the fedvoc meeting of 2025-11-28.
